provider/postgis/postgis.go

```diff
@@ -22,14 +28,121 @@ import (
 
 const Name = "postgis"
 
+type connectionPoolCollector struct {
+	*pgx.ConnPool
+	maxConnectionDesc        *prometheus.Desc
+	currentConnectionsDesc   *prometheus.Desc
+	availableConnectionsDesc *prometheus.Desc
+}
+
+func (c connectionPoolCollector) Describe(ch chan<- *prometheus.Desc) {
+	prometheus.DescribeByCollect(c, ch)
+}
+
+func (c connectionPoolCollector) Collect(ch chan<- prometheus.Metric) {
+	if c.ConnPool == nil {
+		return
+	}
+	stat := c.ConnPool.Stat()
+	ch <- prometheus.MustNewConstMetric(
+		c.maxConnectionDesc,
+		prometheus.GaugeValue,
+		float64(stat.MaxConnections),
+	)
+	ch <- prometheus.MustNewConstMetric(
+		c.currentConnectionsDesc,
+		prometheus.GaugeValue,
+		float64(stat.CurrentConnections),
+	)
+	ch <- prometheus.MustNewConstMetric(
+		c.availableConnectionsDesc,
+		prometheus.GaugeValue,
+		float64(stat.AvailableConnections),
+	)
+}
+
+func (c *connectionPoolCollector) Collectors(prefix string, _ func(configKey string) map[string]interface{}) ([]observability.Collector, error) {
+	if c == nil {
+		return nil, nil
+	}
+	if prefix != "" && !strings.HasSuffix(prefix, "_") {
+		prefix = prefix + "_"
+	}
+
+	c.maxConnectionDesc = prometheus.NewDesc(
+		prefix+"postgres_max_connections",
+		"Max number of postgres connections in the pool",
+		nil,
+		nil,
+	)
+
+	c.currentConnectionsDesc = prometheus.NewDesc(
+		prefix+"postgres_current_connections",
+		"Current number of postgres connections in the pool",
+		nil,
+		nil,
+	)
+
+	c.availableConnectionsDesc = prometheus.NewDesc(
+		prefix+"postgres_available_connections",
+		"Current number of available postgres connections in the pool",
+		nil,
+		nil,
+	)
+	return []observability.Collector{c}, nil
+}
+
 // Provider provides the postgis data provider.
 type Provider struct {
 	config pgx.ConnPoolConfig
-	pool   *pgx.ConnPool
+	pool   *connectionPoolCollector
 	// map of layer name and corresponding sql
 	layers     map[string]Layer
 	srid       uint64
-	firstlayer string
+	firstLayer string
+
+	// registered keeps track if we have already registered these collectors
+	// as the Collectors function will be called for each map and layer, but
+	// we are going to assign those during runtime, instead of at registration
+	// time; so we will only return these collectors on the first call.
+	registered bool
```

Comment:
since this is collector specific, consider renaming to `collectorsRegistred` or similar.